mllib/src/main/scala/org/apache/spark/mllib/linalg/distributed/BlockMatrix.scala

```diff
@@ -0,0 +1,242 @@
+          (this.numPartitions == r.numPartitions)
+      case _ =>
+        false
+    }
+  }
+}
+
+/**
+ * Represents a distributed matrix in blocks of local matrices.
+ *
+ * @param rdd The RDD of SubMatrices (local matrices) that form this matrix
+ * @param nRows Number of rows of this matrix
+ * @param nCols Number of columns of this matrix
+ * @param numRowBlocks Number of blocks that form the rows of this matrix
+ * @param numColBlocks Number of blocks that form the columns of this matrix
+ * @param rowsPerBlock Number of rows that make up each block. The blocks forming the final
+ *                     rows are not required to have the given number of rows
+ * @param colsPerBlock Number of columns that make up each block. The blocks forming the final
+ *                     columns are not required to have the given number of columns
+ */
+class BlockMatrix(
+    val rdd: RDD[((Int, Int), Matrix)],
+    private var nRows: Long,
+    private var nCols: Long,
+    val numRowBlocks: Int,
+    val numColBlocks: Int,
+    val rowsPerBlock: Int,
+    val colsPerBlock: Int) extends DistributedMatrix with Logging {
+
+  private type SubMatrix = ((Int, Int), Matrix) // ((blockRowIndex, blockColIndex), matrix)
+
+  /**
+   * Alternate constructor for BlockMatrix without the input of the number of rows and columns.
+   *
+   * @param rdd The RDD of SubMatrices (local matrices) that form this matrix
+   * @param numRowBlocks Number of blocks that form the rows of this matrix
+   * @param numColBlocks Number of blocks that form the columns of this matrix
+   * @param rowsPerBlock Number of rows that make up each block. The blocks forming the final
+   *                     rows are not required to have the given number of rows
+   * @param colsPerBlock Number of columns that make up each block. The blocks forming the final
+   *                     columns are not required to have the given number of columns
+   */
+  def this(
+      rdd: RDD[((Int, Int), Matrix)],
+      numRowBlocks: Int,
+      numColBlocks: Int,
+      rowsPerBlock: Int,
+      colsPerBlock: Int) = {
+    this(rdd, 0L, 0L, numRowBlocks, numColBlocks, rowsPerBlock, colsPerBlock)
+  }
+
+  private[mllib] var partitioner: GridPartitioner =
+    new GridPartitioner(numRowBlocks, numColBlocks, rdd.partitions.length)
+
+  private lazy val dims: (Long, Long) = getDim
+
+  override def numRows(): Long = {
+    if (nRows <= 0L) nRows = dims._1
+    nRows
+  }
+
+  override def numCols(): Long = {
+    if (nCols <= 0L) nCols = dims._2
+    nCols
+  }
+
+  /** Returns the dimensions of the matrix. */
+  private def getDim: (Long, Long) = {
+    case class MatrixMetaData(var rowIndex: Int, var colIndex: Int,
+        var numRows: Int, var numCols: Int)
+    // picks the sizes of the matrix with the maximum indices
+    def pickSizeByGreaterIndex(example: MatrixMetaData, base: MatrixMetaData): MatrixMetaData = {
```

Comment:
If we modify `base` in-place, we should put `base` as the first parameter. This is the convention used in Spark aggregation.

We should simply this code block. Essentially we want to find the largest row index and col index:

~~~
val (rows, cols) = rdd.map { case (blockRowIndex, blockColIndex, mat) =>
  (blockRowIndex * rowsPerBlock + mat.numRows, blockColIndex * colsPerBlock + mat.numCols)
}.reduce((x0, x1) => (math.max(x0._1, x1._1), math.max(x0._2, x1._2)))
(math.max(rows, nRows), math.max(cols, nCols))
~~~